extra_data/calibration.py

```diff
@@ -0,0 +1,1090 @@
+    CI_GAIN_OF_OF_THRESHOLD = 1 << 4
+    CI_LINEAR_DEVIATION = 1 << 5
+    CI_EVAL_ERROR = 1 << 6
+    FF_GAIN_EVAL_ERROR = 1 << 7
+    FF_GAIN_DEVIATION = 1 << 8
+    FF_NO_ENTRIES = 1 << 9
+    CI2_EVAL_ERROR = 1 << 10
+    VALUE_IS_NAN = 1 << 11
+    VALUE_OUT_OF_RANGE = 1 << 12
+    GAIN_THRESHOLDING_ERROR = 1 << 13
+    DATA_STD_IS_ZERO = 1 << 14
+    ASIC_STD_BELOW_NOISE = 1 << 15
+    INTERPOLATED = 1 << 16
+    NOISY_ADC = 1 << 17
+    OVERSCAN = 1 << 18
+    NON_SENSITIVE = 1 << 19
+    NON_LIN_RESPONSE_REGION = 1 << 20
+    WRONG_GAIN_VALUE = 1 << 21
+
+
+class CCVMetadata(dict):
+    """Dictionary for CCV metadata.
+
+    Identical to a regular dict, but with a custom pandas-based
+    string representation to be easier to read.
+    """
+
+    def __str__(self):
+        """Pretty-print CCV metadata using pandas."""
+
+        import pandas as pd
+
+        res = {pdu_idx: {calibration: ccv_data['ccv_name']
+                         for calibration, ccv_data in pdu_data.items()}
+               for pdu_idx, pdu_data in self.items()}
+
+        return str(pd.DataFrame.from_dict(res, orient='index'))
+
+
+class CalCatError(Exception):
+    """CalCat API error."""
+
+    def __init__(self, response):
+        super().__init__(response['info'])
+
+
+class ClientWrapper(type):
+    """Metaclass to wrap each calibration_client exactly once."""
+
+    _clients = WeakKeyDictionary()
+
+    def __call__(cls, client):
+        instance = cls._clients.get(client, None)
+
+        if instance is None:
+            instance = cls._clients[client] = type.__call__(cls, client)
+
+        return instance
+
+
+class CalCatApi(metaclass=ClientWrapper):
+    """Internal calibration_client wrapper."""
+
+    get_detector_keys = ['id', 'name', 'identifier', 'karabo_name',
+                         'karabo_id_control', 'description']
+    get_pdu_keys = ['id', 'physical_name', 'karabo_da', 'virtual_device_name',
+                    'detector_type_id', 'detector_id', 'description']
+
+    def __init__(self, client):
+        self.client = client
+
+    @classmethod
+    def format_time(cls, dt):
+        """Parse different ways to specify time to CalCat."""
+
+        if isinstance(dt, datetime):
+            return dt.astimezone(timezone.utc).strftime('%Y%m%dT%H%M%S%Z')
+        elif isinstance(dt, date):
+            return cls.format_time(datetime.combine(dt, time()))
+
+        return dt
+
+    def format_cond(self, condition):
+        """Encode operating condition to CalCat API format.
+
+        Args:
+            caldata (CalibrationData): Calibration data instance used to
+                interface with database.
+
+        Returns:
+            (dict) Operating condition for use in CalCat API.
+        """
+
+        return {'parameters_conditions_attributes': [
+            {'parameter_id': self.parameter_id(k), 'value': str(v)}
+            for k, v in condition.items()]}
+
+    @lru_cache()
+    def detector(self, detector_name):
```

Comment:
I came across [a blog post](https://rednafi.github.io/reflections/dont-wrap-instance-methods-with-functoolslru_cache-decorator-in-python.html) the other day pointing out that using `lru_cache` on regular methods like this prevents instances from being freed - the cache is kept on the class, and keeps references to each instance where the method has been called. The suggested workaround was to wrap the methods in `__init__`:

```python
self.calculate = lru_cache()(self._calculate)
```

Probably not a big deal here, because we're unlikely to create thousands of `CalCatApi` instances, but it might be best to avoid the issue anyway, in case we change how the code works or copy this pattern somewhere else.